inst/include/model_simulation.hpp

```diff
@@ -0,0 +1,135 @@
+#pragma once
+
+#include "types.hpp"
+
+namespace leapfrog {
+
+namespace internal {
+
+template<typename real_type>
+void run_hiv_model_simulation(int time_step,
+                              const Parameters<real_type> &pars,
+                              const State<real_type> &state_curr,
+                              State<real_type> &state_next,
+                              IntermediateData<real_type> &intermediate) {
+  run_add_new_hiv_infections(time_step, pars, state_curr, state_next, intermediate);
+
+  intermediate.everARTelig_idx =
+      pars.artcd4elig_idx(time_step) < pars.disease_stages ? pars.artcd4elig_idx(time_step) : pars.disease_stages;
+  for (int hiv_step = 0; hiv_step < pars.hiv_steps_per_year; ++hiv_step) {
+    run_disease_progression_and_mortality(time_step, pars, state_curr, state_next, intermediate, hiv_step);
+    run_new_infections(time_step, pars, state_curr, state_next, intermediate, hiv_step);
+  }
+}
+
+template<typename real_type>
+void run_add_new_hiv_infections(int time_step,
+                                const Parameters<real_type> &pars,
+                                const State<real_type> &state_curr,
+                                State<real_type> &state_next,
+                                IntermediateData<real_type> &intermediate) {
+  // TODO: Add EPP_DIRECTINCID switch
+
+  // Calculating new infections once per year (like Spectrum)
+  for (int g = 0; g < pars.num_genders; ++g) {
+    for (int a = pars.adult_incidence_first_age_group; a < pars.age_groups_pop; a++) {
+      intermediate.hiv_negative_pop(a, g) = state_curr.total_population(a, g) - state_curr.hiv_population(a, g);
+    }
+  }
+
+  for (int g = 0; g < pars.num_genders; g++) {
+    for (int a = pars.adult_incidence_first_age_group;
+         a < pars.adult_incidence_first_age_group + pars.pAG_INCIDPOP; a++) {
+      intermediate.hiv_neg_aggregate(g) += intermediate.hiv_negative_pop(a, g);
+      intermediate.Xhivn_incagerr(g) +=
+          pars.incidence_relative_risk_age(a - pars.adult_incidence_first_age_group, g, time_step) *
+          intermediate.hiv_negative_pop(a, g);
+    }
+  }
+
+  intermediate.incidence_rate_sex(MALE) =
+      pars.incidence_rate(time_step) * (intermediate.hiv_neg_aggregate(MALE) + intermediate.hiv_neg_aggregate(FEMALE)) /
+      (intermediate.hiv_neg_aggregate(MALE) +
+       pars.incidence_relative_risk_sex(time_step) * intermediate.hiv_neg_aggregate(FEMALE));
+  intermediate.incidence_rate_sex(FEMALE) =
+      pars.incidence_rate(time_step) * pars.incidence_relative_risk_sex(time_step) *
+      (intermediate.hiv_neg_aggregate(MALE) + intermediate.hiv_neg_aggregate(FEMALE)) /
+      (intermediate.hiv_neg_aggregate(MALE) +
+       pars.incidence_relative_risk_sex(time_step) * intermediate.hiv_neg_aggregate(FEMALE));
+
+  for (int g = 0; g < pars.num_genders; g++) {
```

Comment:
Check with Jeff why some of the gender stuff is looped and some is explicit

Reply:
Actually here is ebcause above the formula for male and female isn't the same, we have

for male
```
incidence_rate_male  = incidence_rate * (hiv_neg_male + hiv_neg_female) / (hiv_neg_male + incidence_rate_relative_risk * hiv_neg_female)
```

for female

```
incidence_rate_female = incidence_rate * incidence_rate_relative_risk * (hiv_neg_male + hiv_neg_female) / (hiv_neg_male + incidence_rate_relative_risk * hiv_neg_female)
```